legate/core/communicator.py

```diff
@@ -104,3 +104,51 @@ def _finalize(self, volume: int, handle: FutureMap) -> None:
         task = Task(self._context, self._finalize_nccl, tag=self._tag)
         task.add_future_map(handle)
         task.execute(Rect([volume]))
+
+
+class CPUCommunicator(Communicator):
+    def __init__(self, runtime):
+        super(CPUCommunicator, self).__init__(runtime)
+        library = runtime.core_library
+
+        self._init_cpucoll_id = library.LEGATE_CORE_INIT_CPUCOLL_ID_TASK_ID
+        self._init_cpucoll_mapping = (
+            library.LEGATE_CORE_INIT_CPUCOLL_MAPPING_TASK_ID
+        )
+        self._init_cpucoll = library.LEGATE_CORE_INIT_CPUCOLL_TASK_ID
+        self._finalize_cpucoll = library.LEGATE_CORE_FINALIZE_CPUCOLL_TASK_ID
+        self._tag = library.LEGATE_CPU_VARIANT
+        self._needs_barrier = False
+
+    def destroy(self) -> None:
+        Communicator.destroy(self)
+        self._runtime.issue_execution_fence(block=True)
```

Comment:
Two comments: 1) please leave a comment why we need this blocking. 2) do this only if you have any CPU communicators. many programs don't ever create them.

Reply:
This is the destructor of communicator, so if programs do not create them, the destroy function will not be called, right?

Reply:
`Communicator.destroy` is invoked unconditionally once per communicator type. it's `Communicator._finalize` that is invoked for each created communicator. the fence is to block on the tasks launched by those `Communicator._finalize` calls, so you don't need it if no communicator is created. I assume you still need to `legate_cpucoll_finalize` even when no CPU communicator is created. but I believe that function can be invoked without the fence in that case.

Reply:
Got it, I have pushed a fix for it. 
https://github.com/nv-legate/legate.core/pull/214/commits/d319db6326ce89f4c604c0f2236af44b3cbce6c0